Rest api challenge

jhalak04 opened the pull request:
Please see below file for code - 

[Steps to run application on local](https://github.com/Statflo/challenge-rest-api-java/blob/4c38b93d4b15bab710bd98f9ede3e1fef2a881be/sample_app/READMEAPI.md)

[Review controller code](https://github.com/jhalak04/challenge-rest-api-java/blob/461e4e3b44cefe92be4d43bddc221ce84e53b773/sample_app/app/Http/Controllers/UserDetailsController.php)

[Review PHPUnit test code](https://github.com/jhalak04/challenge-rest-api-java/blob/461e4e3b44cefe92be4d43bddc221ce84e53b773/sample_app/tests/Unit/UserDetailsTest.php)

kinncj commented:
Thanks for the submission,
our team will be analyzing the challenge in order to continue with the interview process.

So far, that's my review:

**UserDetailsController::index()**

1. I've updated the method so we can apply the filters across all records at the SQL level instead of the Collection level.
2. After the update, it will bring all matches instead of a single one.
3. After the update, it will compose all possible filters instead of returning on the first one.

_Considerations:_

- Use a service layer between the Repository and the Controller, so it can be decoupled and the Controller doesn't need to be aware of the Domain in specific.

    
**.env.sample**

1. Updated with a sample APP_KEY, otherwise PHPUnit fails. The reason why it fails will be covered under the GENERAL statements

_Considerations:_

- Always add all required values to the sample file, even if those values are dummy values, so it's easier to identify what's requires and what's not.

**Bonus:**

1. Added Docker so we have a unified dev environment
2. Moved the project to the root folder, so the CI tool can run the unit tehttps://github.com/jhalak04/challenge-rest-api-java/tree/sample_appsts

    
**GENERAL:**

- Laravel violates a couple [S.O.L.I.D](https://en.wikipedia.org/wiki/SOLID) principles on its own... the reason why UNIT testing is so difficult, is due to the lack of `Dependency Injection` and highly coupled with all those static methods...
- Given that, some of the UNIT tests were failing due to the lack of ability to mock some of the framework's internal calls, such as the Encryption portion of it.

_Useful links:_

- https://www.linkedin.com/pulse/writing-solid-laravel-code-andy-beak/
- https://medium.com/devcupboard/a-different-view-on-dependency-injection-in-laravel-c01ec690d9ee

I've created a `patch` file for you, so you can see those changes locally.
https://gist.github.com/kinncj/00bd04b526710c39d7a60cf3c4550a81

The patch can be applied with `git apply challenge_rest_api_java_PR_7.patch`

jhalak04 commented:
Hello Kinn,

Thanks for reviewing pull request. I will definitely check given patch and
look into the improvement you suggest.

Looking forward to hearing from your team soon.

Thank you.
Jhalak.


On Mon., Apr. 1, 2019, 12:38 Kinn Coelho Juliao, <notifications@github.com>
wrote:

> Thanks for the submission,
> our team will be analyzing the challenge in order to continue with the
> interview process.
>
> So far, that's my review:
>
> *UserDetailsController::index()*
>
>    1. I've updated the method so we can apply the filters across all
>    records at the SQL level instead of the Collection level.
>    2. After the update, it will bring all matches instead of a single one.
>    3. After the update, it will compose all possible filters instead of
>    returning on the first one.
>
> *Considerations:*
>
>    - Use a service layer between the Repository and the Controller, so it
>    can be decoupled and the Controller doesn't need to be aware of the Domain
>    in specific.
>
> *.env.sample*
>
>    1. Updated with a sample APP_KEY, otherwise PHPUnit fails. The reason
>    why it fails will be covered under the GENERAL statements
>
> *Considerations:*
>
>    - Always add all required values to the sample file, even if those
>    values are dummy values, so it's easier to identify what's requires and
>    what's not.
>
> *Bonus:*
>
>    1. Added Docker so we have a unified dev environment
>    2. Moved the project to the root folder, so the CI tool can run the
>    unit tehttps://
>    github.com/jhalak04/challenge-rest-api-java/tree/sample_appsts
>
> *GENERAL:*
>
>    - Laravel violates a couple S.O.L.I.D
>    <https://en.wikipedia.org/wiki/SOLID> principles on its own... the
>    reason why UNIT testing is so difficult, is due to the lack of Dependency
>    Injection and highly coupled with all those static methods...
>    - Given that, some of the UNIT tests were failing due to the lack of
>    ability to mock some of the framework's internal calls, such as the
>    Encryption portion of it.
>
> *Useful links:*
>
>    - https://www.linkedin.com/pulse/writing-solid-laravel-code-andy-beak/
>    -
>    https://medium.com/devcupboard/a-different-view-on-dependency-injection-in-laravel-c01ec690d9ee
>
> I've created a patch file for you, so you can see those changes locally.
> https://gist.github.com/kinncj/00bd04b526710c39d7a60cf3c4550a81
>
> The patch can be applied with git apply challenge_rest_api_java_PR_7.patch
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Statflo/challenge-rest-api-java/pull/7#issuecomment-478653322>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AcFJ-UcQlk2Ay9NKjhGkMvpoxOdxPEptks5vcjXzgaJpZM4cUFsA>
> .
>
